src/kv/RegionCache.cc

```diff
@@ -238,14 +242,30 @@ Store RegionCache::getStore(Backoffer & bo, uint64_t id)
     return reloadStore(store);
 }
 
-std::pair<std::vector<uint64_t>, std::vector<uint64_t>> RegionCache::getAllValidTiFlashStores(Backoffer & bo, const RegionVerID & region_id, const Store & current_store, const LabelFilter & label_filter)
+std::pair<std::vector<uint64_t>, std::vector<uint64_t>> RegionCache::getAllValidTiFlashStores(Backoffer & bo, const RegionVerID & region_id, const Store & current_store, const LabelFilter & label_filter, const std::unordered_set<uint64_t> * store_id_blocklist)
 {
+    auto remove_blocklist = [](const std::unordered_set<uint64_t> * store_id_blocklist, std::vector<uint64_t> & stores, const RegionVerID & region_id, Logger * log) {
+        if (store_id_blocklist != nullptr)
+        {
+            auto origin_size = stores.size();
+            stores.erase(std::remove_if(stores.begin(), stores.end(), [&](int x) {
+                             return store_id_blocklist->find(x) != store_id_blocklist->end();
+                         }),
+                         stores.end());
+            if (log != nullptr && origin_size != stores.size())
+            {
+                auto s = "blocklist peer removed, region=" + region_id.toString() + ", origin_store_size=" + std::to_string(origin_size) + ", current_store_size=" + std::to_string(stores.size());
+                log->information(s);
+            }
+        }
+    };
     std::vector<uint64_t> all_stores;
     std::vector<uint64_t> non_pending_stores;
     RegionPtr cached_region = getRegionByIDFromCache(region_id);
     if (cached_region == nullptr)
     {
         all_stores.emplace_back(current_store.id);
+        remove_blocklist(store_id_blocklist, all_stores, region_id, log);
```

Comment:
`balanceBatchCopTasks` may meet out-of-index in this line.
https://github.com/tikv/client-c/blob/617d23301a6757238bdc5a6eca0b5cc87b680f43/src/coprocessor/Client.cc#L158

Reply:
In buildBatchCopTasks, if no store can be selected, it won't go that far. It will throw an exception in backoffer because `getRPCContext` returns none.